### PR TITLE
Optics correction

### PR DESCRIPTION
**Pull request by murilobalves**
- LOCO-like optics correction
- All quadrupoles used individually as default
- Additional and proportional methods
- The code minimizes the difference of residue vector which includes betas at BPMs e correctors (betax@CH and betay@CV), horizontal dispersion at BPMs and tunes.
- It is intended that in the next version this code will include optics symmetrization based on the code implemented in AT.

### CODE REVIEW

File: apsuite/optics_analysis/optics_correction.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        if model is None:
+            model = self.model
+        if knobsidx is None:
+            knobsidx = self.knobs_idx
+        kl = []
+        for mag in knobsidx:
+            kl_seg = []
+            for seg in mag:
+                kl_seg.append(model[seg].KL)
+            kl.append(kl_seg)
```
Comment (fernandohds564) [inline apsuite/optics_analysis/optics_correction.py]: I think you have to sum the KL of the segments here, don't you?

File: apsuite/optics_analysis/optics_correction.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        return np.array(kl)
```
Comment (fernandohds564) [inline apsuite/optics_analysis/optics_correction.py]: If you don't sum the KL as commented above. Here you are assuming all knobs have the same number of segments. I know that for the current version of Sirius lattice this is true, but should we make this assumption here?

File: apsuite/optics_analysis/optics_correction.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        res0 = self._get_optics_vector(model)
+        self.jacobian_matrix = np.zeros((res0.size, len(self.knobs_idx)))
+        delta = 1e-6
+
+        modcopy = model[:]
+        for idx, nmag in enumerate(self.knobs_idx):
+            dlt = delta/len(nmag)
+            for seg in nmag:
+                modcopy[seg].KL += dlt
```
Comment (fernandohds564) [inline apsuite/optics_analysis/optics_correction.py]: with this implementation, the total `KL` you are applying to each element is `n*dlt` where `n` is the number of segments of the element. Is really that what you want?

File: apsuite/optics_analysis/optics_correction.py (hunk 1 above)
Comment (murilobalves) [reply apsuite/optics_analysis/optics_correction.py]: yes! done.

File: apsuite/optics_analysis/optics_correction.py (hunk 2 above)
Comment (murilobalves) [reply apsuite/optics_analysis/optics_correction.py]: fixed!

File: apsuite/optics_analysis/optics_correction.py (hunk 3 above)
Comment (murilobalves) [reply apsuite/optics_analysis/optics_correction.py]: yes! 'dlt = delta/len(nmag)' is the division of the total delta in 'N' parts where 'N' is the number of segments. Then I apply 'dlt' for each one of the 'N' segments, therefore obtaining the total 'delta' for each element as desired.

File: apsuite/optics_analysis/optics_correction.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        else:
+            return OpticsCorr.CORR_STATUS.Fail
+        return OpticsCorr.CORR_STATUS.Sucess
+
+    def optics_correction(self,
+                          model,
+                          goal_model=None,
+                          jacobian_matrix=None,
+                          nsv=None, nr_max=10, tol=1e-6,
+                          method='loco'):
```
Comment (fernandohds564) [inline apsuite/optics_analysis/optics_correction.py]: I think this `method`option should be an object property, not an input argument.

Comment (murilobalves) [reply apsuite/optics_analysis/optics_correction.py]: done.

File: apsuite/optics_analysis/optics_correction.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    METHODS = _namedtuple('Methods', ['Additional', 'Proportional'])(0, 1)
+    CORR_STATUS = _namedtuple('CorrStatus', ['Fail', 'Sucess'])(0, 1)
+
+    def __init__(self, model, acc, dim='4d', knobs_list=None,
+                 method=None, correction_method='loco'):
+        """."""
+        self.model = model
+        self.acc = acc
+        self.dim = dim
+        self.corr_method = correction_method
```
Comment (fernandohds564) [inline apsuite/optics_analysis/optics_correction.py]: treat it as a hidden attribute with a property, just like the `method` attribute.

File: apsuite/optics_analysis/optics_correction.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                          jacobian_matrix=None,
+                          nsv=None, nr_max=10, tol=1e-6):
+        """Optics correction method selection.
+
+        Methods available:
+        - LOCO-like correction
+        """
+        if model is None:
+            model = self.model
+        if self.corr_method == 'loco':
```
Comment (fernandohds564) [inline apsuite/optics_analysis/optics_correction.py]: here, the comparison should be against the `namedtuple` you will define.

File: apsuite/optics_analysis/optics_correction.py (hunk 4 above)
Comment (fernandohds564) [reply apsuite/optics_analysis/optics_correction.py]: see comments below.

**Pull request merged**